src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointProvider.cs

```diff
@@ -53,43 +43,9 @@ internal class ServiceEndpointProvider : IServiceEndpointProvider, IDisposable
 
             _generator = new DefaultServiceEndpointGenerator(endpoint);
 
-            _ = UpdateAccessKeyAsync(provider, endpoint, loggerFactory ?? NullLoggerFactory.Instance);
-        }
-
-        public async Task UpdateAccessKeyAsync(IServerNameProvider provider, ServiceEndpoint endpoint, ILoggerFactory loggerFactory)
-        {
             if (endpoint.AccessKey is AadAccessKey key)
-                _timer.Start();
-            else
-                return;
-
-            var logger = loggerFactory.CreateLogger<ServiceEndpointProvider>();
-
-            while (await _timer)
             {
-                var isAuthorized = false;
-                for (int i = 0; i < AuthorizeMaxRetryTimes; i++)
-                {
-                    logger.LogInformation($"Try authorizing AccessKey...({i})");
-                    var source = new CancellationTokenSource(AuthorizeTimeout);
-                    try
-                    {
-                        await key.AuthorizeAsync(endpoint.Endpoint, endpoint.Port, provider.GetName(), source.Token);
-                        logger.LogInformation("AccessKey has been authorized successfully.");
-                        isAuthorized = true;
-                        break;
-                    }
-                    catch (Exception e)
-                    {
-                        logger.LogWarning(e, $"Failed to authorize AccessKey, will retry in {AuthorizeRetryIntervalInSec} seconds.");
-                        await Task.Delay(AuthorizeRetryInterval);
-                    }
-                }
-
-                if (!isAuthorized)
-                {
-                    logger.LogError($"AccessKey authorized failed more than {AuthorizeMaxRetryTimes} times.");
-                }
+                _ = key.UpdateAccessKeyAsync(provider, endpoint, loggerFactory ?? NullLoggerFactory.Instance);
```

Comment:
shall the accessKey itself control when to start updating the key?

Reply:
In current design, provider is created for every request, it should be super lightweighted

Reply:
it's already super lightweight, only the first call will be executed, the method is protected by a TaskCompletionSource.